Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,395 @@
+  where
+    proof := by
+      rw [size_extract, v.size_eq]
+
+/-- Custom eliminator for `Vector α n` through Arrays -/
+@[elab_as_elim]
+def elimAsArray {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : Array α, motive ⟨a, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨a, rfl⟩ => mk a
+
+/-- Custom eliminator for `Vector α n` throughLists -/
+@[elab_as_elim]
+def elimAsList {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : List α, motive ⟨⟨a⟩, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨⟨a⟩, rfl⟩ => mk a
+
+/-- Map a vector under a function. -/
+def map (f : α → β) : Vector α n → Vector β n
+  | ⟨a, h⟩ => ⟨Array.map f a, by simp [*]⟩
+
+/-- Mapping two vectors under a curried function of two variables. -/
+def zipWith : Vector α n → Vector β n → (α → β → φ) → Vector φ n
+  | ⟨a, h₁⟩, ⟨b, h₂⟩, f => ⟨Array.zipWith a b f, by simp [Array.size_zipWith, h₁, h₂]⟩
+
+/-- Drop `i` elements from a vector of length `n`; we can have `i > n`. -/
+def drop (i : Nat) (v : Vector α n) : Vector α (n - i) :=
+  have : min n n - i = n - i := by
+    rw [Nat.min_self]
+  Vector.cast this (extract v i n)
+
+/-- Take `i` elements from a vector of length `n`; we can have `i > n`. -/
+def take (i : Nat) (v : Vector α n) : Vector α (min i n) :=
+  extract v 0 i
+
+/-- Vector of length `n` from a function on `Fin n`. -/
+def ofFn (f : Fin n → α) : Vector α n := ⟨Array.ofFn f, by {rw [size_ofFn]}⟩
+
+/--
+Swaps two entries in a Vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swap (v : Vector α n) (i j : Fin n) : Vector α n :=
+  ⟨Array.swap v.toArray (Fin.cast v.size_eq.symm i) (Fin.cast v.size_eq.symm j), proof⟩
+  where
+    proof := by
+      rw [size_swap]
+      exact v.size_eq
+
+/--
+Swaps two `Nat` indexed entries in a `Vector α n`.
+Uses `get_elem_tactic` to supply a proof that the indices are in range.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapN (v : Vector α n) (i j : Nat)
+    (h_i : i < n := by get_elem_tactic) (h_j : j < n := by get_elem_tactic) : Vector α n :=
+  ⟨Array.swap v.toArray (Fin.cast v.size_eq.symm ⟨i, h_i⟩) (Fin.cast v.size_eq.symm ⟨j, h_j⟩),
+    proof⟩
+  where
+    proof := by
+      rw [size_swap]
+      exact v.size_eq
+
+/--
+Swaps two entries in a `Vector α n`, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swap! (v : Vector α n) (i j : Nat) : Vector α n :=
+  ⟨Array.swap! v.toArray i j, by {rw [size_swap!]; exact v.size_eq}⟩
+
+/--
+Swaps the entry with index `i : Fin n` in the vector for a new entry.
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt (v : Vector α n) (i : Fin n) (x : α) : α × Vector α n:=
```

Comment:
I'm not sure this actually does destructive updates. The array code doesn't make that claim.

It would be better to call `Array.swapAt` since that is more likely to be fixed so that destructive updates do happen.

Reply:
Done.